### ISSUE TITLE
[Phase 2] Basic Multi-Label Classification

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/multi-label-doc-classifier#3]
## Summary

Implement core multi-label classification using classifier chains and label powerset approaches with comprehensive evaluation metrics.

**Parent Issue**: #1
**Depends On**: Phase 1 (Data Foundation)

## Objectives

- Implement classifier chain for capturing label dependencies
- Implement label powerset with hashing for tractable multi-label handling
- Create multi-label evaluation metrics (Hamming loss, F1, etc.)
- Add cross-validation support for model selection

## System Context

```
src/
└── models/
    ├── base_classifier.py     # NEW - Abstract base class
    ├── classifier_chain.py    # NEW - Classifier chain implementation
    └── label_powerset.py      # NEW - Label powerset with hashing

src/
└── evaluation/
    └── metrics.py             # NEW - Multi-label metrics
```

### Base Classifier
- [ ] Create `src/models/base_classifier.py`
- [ ] Define abstract `MultiLabelClassifier` interface
- [ ] Methods: `fit()`, `predict()`, `predict_proba()`
- [ ] Support sklearn-compatible estimators

### Classifier Chain
- [ ] Create `src/models/classifier_chain.py`
- [ ] Implement `ClassifierChain` class
- [ ] Chain labels in order, using previous predictions as features
- [ ] Support configurable base estimator (LogisticRegression default)
- [ ] Add random label ordering option for ensembling

### Label Powerset
- [ ] Create `src/models/label_powerset.py`
- [ ] Implement `LabelPowersetClassifier` class
- [ ] Use feature hashing for high-cardinality label combinations
- [ ] Handle unseen label combinations in prediction

### Evaluation Metrics
- [ ] Create `src/evaluation/metrics.py`
- [ ] Implement: `hamming_loss`, `subset_accuracy`, `micro_f1`, `macro_f1`
- [ ] Per-label precision, recall, F1 scores
- [ ] Classification report generator

### ClassifierChain Skeleton

```python
# src/models/classifier_chain.py
import numpy as np
from sklearn.base import clone
from sklearn.linear_model import LogisticRegression
from typing import Optional
from scipy.sparse import hstack, csr_matrix

class ClassifierChain:
    """Classifier chain for multi-label classification.

    Trains a sequence of binary classifiers where each classifier
    uses the predictions of all previous classifiers as additional features.
    """

    def __init__(
        self,
        base_estimator=None,
        order: Optional[list] = None,
        random_state: Optional[int] = None
    ):
        self.base_estimator = base_estimator or LogisticRegression(max_iter=1000)
        self.order = order
        self.random_state = random_state
        self.classifiers_ = []
        self.n_labels_ = 0

    def fit(self, X, y):
        """Fit the classifier chain.

        Args:
            X: Feature matrix (n_samples, n_features)
            y: Label matrix (n_samples, n_labels)
        """
        y = np.asarray(y)
        self.n_labels_ = y.shape[1]

        # Determine label order
        if self.order is None:
            self.order_ = list(range(self.n_labels_))
        else:
            self.order_ = list(self.order)

        # Train classifiers in chain order
        self.classifiers_ = []
        X_chain = X

        for i, label_idx in enumerate(self.order_):
            clf = clone(self.base_estimator)

            # Add previous predictions as features
            if i > 0:
                prev_preds = y[:, self.order_[:i]]
                if hasattr(X_chain, 'toarray'):
                    X_augmented = hstack([X, csr_matrix(prev_preds)])
                else:
                    X_augmented = np.hstack([X, prev_preds])
            else:
                X_augmented = X

            clf.fit(X_augmented, y[:, label_idx])
            self.classifiers_.append(clf)

        return self

    def predict(self, X):
        """Predict labels for samples.

        Args:
            X: Feature matrix (n_samples, n_features)

        Returns:
            Predicted label matrix (n_samples, n_labels)
        """
        n_samples = X.shape[0]
        predictions = np.zeros((n_samples, self.n_labels_))

        for i, label_idx in enumerate(self.order_):
            clf = self.classifiers_[i]

            if i > 0:
                prev_preds = predictions[:, self.order_[:i]]
                if hasattr(X, 'toarray'):
                    X_augmented = hstack([X, csr_matrix(prev_preds)])
                else:
                    X_augmented = np.hstack([X, prev_preds])
            else:
                X_augmented = X

            predictions[:, label_idx] = clf.predict(X_augmented)

        return predictions
```

### Multi-Label Metrics

```python
# src/evaluation/metrics.py
import numpy as np
from typing import Dict

def hamming_loss(y_true: np.ndarray, y_pred: np.ndarray) -> float:
    """Fraction of labels that are incorrectly predicted."""
    y_true = np.asarray(y_true)
    y_pred = np.asarray(y_pred)
    return np.mean(y_true != y_pred)

def subset_accuracy(y_true: np.ndarray, y_pred: np.ndarray) -> float:
    """Fraction of samples with exactly correct label set."""
    y_true = np.asarray(y_true)
    y_pred = np.asarray(y_pred)
    return np.mean(np.all(y_true == y_pred, axis=1))

def micro_f1(y_true: np.ndarray, y_pred: np.ndarray) -> float:
    """F1 score computed globally across all labels."""
    y_true = np.asarray(y_true).flatten()
    y_pred = np.asarray(y_pred).flatten()

    tp = np.sum((y_true == 1) & (y_pred == 1))
    fp = np.sum((y_true == 0) & (y_pred == 1))
    fn = np.sum((y_true == 1) & (y_pred == 0))

    precision = tp / (tp + fp) if (tp + fp) > 0 else 0
    recall = tp / (tp + fn) if (tp + fn) > 0 else 0

    if precision + recall == 0:
        return 0
    return 2 * precision * recall / (precision + recall)

def multilabel_report(y_true: np.ndarray, y_pred: np.ndarray, label_names: list = None) -> Dict:
    """Generate comprehensive multi-label classification report."""
    return {
        'hamming_loss': hamming_loss(y_true, y_pred),
        'subset_accuracy': subset_accuracy(y_true, y_pred),
        'micro_f1': micro_f1(y_true, y_pred),
    }
```

## Verification

```python
from sklearn.model_selection import train_test_split
from src.models.classifier_chain import ClassifierChain
from src.evaluation.metrics import hamming_loss, subset_accuracy

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Train classifier chain
chain = ClassifierChain()
chain.fit(X_train, y_train)

# Evaluate
y_pred = chain.predict(X_test)
print(f"Hamming Loss: {hamming_loss(y_test, y_pred):.4f}")
print(f"Subset Accuracy: {subset_accuracy(y_test, y_pred):.4f}")

assert y_pred.shape == y_test.shape, "Prediction shape should match ground truth"
assert hamming_loss(y_test, y_pred) < 0.5, "Should beat random baseline"
```

## Definition of Done

- [ ] Classifier chain trains and predicts without errors
- [ ] Label powerset handles 50+ unique label combinations
- [ ] All metrics compute correctly on synthetic test data
- [ ] Cross-validation runs successfully with 5-fold CV
- [ ] Unit tests achieve 80% code coverage

## Files to Modify

| File | Lines | Action | Description |
|------|-------|--------|-------------|
| `src/models/base_classifier.py` | - | Create | Abstract base class |
| `src/models/classifier_chain.py` | - | Create | Classifier chain |
| `src/models/label_powerset.py` | - | Create | Label powerset |
| `src/evaluation/metrics.py` | - | Create | Multi-label metrics |
| `tests/test_classifier_chain.py` | - | Create | Chain unit tests |
| `tests/test_metrics.py` | - | Create | Metrics unit tests |
